server/src/main/java/io/druid/server/lookup/cache/LookupCoordinatorManager.java

```diff
@@ -506,137 +265,441 @@ public boolean deleteLookup(final String tier, final String lookup, AuditInfo au
+  {
+    LOG.debug("Starting lookup sync for node [%s].", node);
+
+    LookupsStateWithMap currLookupsStateOnNode = getLookupStateForNode(node);
+    LOG.debug("Received lookups state from node [%s].", node);
+
+
+    // Compare currLookupsStateOnNode with nodeTierLookupsToBe to find what are the lookups
+    // we need to further ask node to load/drop
+    Map<String, LookupExtractorFactoryMapContainer> toLoad = getToBeLoadedOnNode(
+        currLookupsStateOnNode,
+        nodeTierLookupsToBe
+    );
+    Set<String> toDrop = getToBeDroppedFromNode(currLookupsStateOnNode, nodeTierLookupsToBe);
+
+    if (!toLoad.isEmpty() || !toDrop.isEmpty()) {
+      // Send POST request to the node asking to load and drop the lookups necessary.
+      // no need to send "current" in the LookupsStateWithMap , that is not required
+      currLookupsStateOnNode = updateNode(node, new LookupsStateWithMap(null, toLoad, toDrop));
+
+      LOG.debug(
+          "Sent lookup toAdd[%s] and toDrop[%s] updates to node [%s].",
+          toLoad.keySet(),
+          toDrop,
+          node
+      );
+    }
+
+    LOG.debug("Finished lookup sync for node [%s].", node);
+    return currLookupsStateOnNode;
+  }
+
+  // Returns the Map<lookup-name, lookup-spec> that needs to be loaded by the node and it does not know about
+  // those already.
+  private Map<String, LookupExtractorFactoryMapContainer> getToBeLoadedOnNode(
+      LookupsStateWithMap currLookupsStateOnNode,
+      Map<String, LookupExtractorFactoryMapContainer> nodeTierLookupsToBe
+  )
+  {
+    Map<String, LookupExtractorFactoryMapContainer> toLoad = new HashMap<>();
+    for (Map.Entry<String, LookupExtractorFactoryMapContainer> e : nodeTierLookupsToBe.entrySet()) {
+      String name = e.getKey();
+      LookupExtractorFactoryMapContainer lookupToBe = e.getValue();
+
+      // get it from the current pending notices list on the node
+      LookupExtractorFactoryMapContainer current = currLookupsStateOnNode.getToLoad().get(name);
+
+      if (current == null) {
+        //ok, not on pending list, get from currently loaded lookups on node
+        current = currLookupsStateOnNode.getCurrent().get(name);
+      }
+
+      if (current == null || //lookup is neither pending nor already loaded on the node OR
+          lookupToBe.replaces(current) //lookup is already know to node, but lookupToBe overrides that
+          ) {
+        toLoad.put(name, lookupToBe);
+      }
+    }
+    return toLoad;
+  }
+
+  // Returns Set<lookup-name> that should be dropped from the node which has them already either in pending to load
+  // state or loaded
+  private Set<String> getToBeDroppedFromNode(
+      LookupsStateWithMap currLookupsStateOnNode,
+      Map<String, LookupExtractorFactoryMapContainer> nodeTierLookupsToBe
+  )
+  {
+    Set<String> toDrop = new HashSet<>();
+
+    // {currently loading/loaded on the node} - {currently pending deletion on node} - {lookups node should actually have}
```

Comment:
This logic doesn't manage the order of toDrop/toLoad notices, which are already on the node. Well, as it is implemented currently, it seems that those lists are always disjoint (though I'm not sure), but this is a subtle and undocumented property. Suggested to change  `LookupsStateWithMap` and `LookupsState` so that it has `current` and `pending` maps. In this case disjoint is ensured by design.